### DISCUSSION TITLE
GrandOrgue and Jack on Ubuntu 22.04

### DISCUSSION THREAD

**Discussion by thbeckpr** [GrandOrgue/grandorgue#1387] [Help]
I'm running GO v3.9.5-1 on Ubuntu xfce 22.04.

I have a large sample set (Sonus Paradisi Buckeburg demo) that plays poorly using Portaudio. There are many xruns and it is basically unusable. I decided to try using Jack and it works well. However, for some reason, I can no longer get sound using jack. I'm not sure what changed. I certainly didn't intentionally change anything.

I'm starting jack with this command: /usr/bin/jackd -dalsa -dhw:PCH -r48000 -p196 -n2

The GO audio settings are:
sample rate = 48000
samples per buffer = 196

The selected sound port is Jack and the audio output device is Jack Native Output.

Any ideas?

**Comment by oleg68** [comment]
1. samples per buffer should be 128 or 256, but not 192
2. When using Jack Native Output you have to connect GrandOrgueAudio jack ports to `system` or any other input ports manually. Using `qjackctl` is recommended. It also allows to automate these connections.

**Comment by thbeckpr** [comment]
It does work if I start from qjackctl and set the audio connections.
The samples per buffer do not have to be a power of 2. A setting of 128 causes xruns, but 256 causes noticeable latency.

**Reply by oleg68** [reply]
> A setting of 128 causes xruns, but 256 causes noticeable latency

Tty to increase the number of buffers, ex. ``jackd -dalsa -dhw:PCH -r48000 -p128 -n5``

**Comment by thbeckpr** [comment]
I read in the jackd documentation that the samples per buffer must be a power of 2 as you say. So I switched to -p256 -n5 and still had xruns. I switched back to -p192 -n2 with no xruns and good sound.

So I'm not sure why samples per buffer need to be a power of 2 but  192 has solved my problem so I'm going to keep it.

**Reply by thbeckpr** [reply]
I should have added, I don't know how to make the audio connections using the cli so I'm using qjackctl.

**Reply by oleg68** [reply]
Tthere are two ways for automating jack connections:
- make a `Patchbay` in qjackctl, save it in a profile and run qjackctl with the profile name from a shell script
- use `jack_connect` command from a shell script

**Reply by thbeckpr** [reply]
Thanks Oleg. I'm going to start qjackctl in a script followed by my app. It's working great.

The patch can also be activated on startup in the setup options of qjackctl.

**Reply by thbeckpr** [reply]
How do I mark Oleg's comment as the answer?

**Comment by midimusic** [comment]
Please explain what an "xrun" is. I do not know how jack can "improve" sound quality. All it does is add extra sound processing between the source of sound, GO  and its destination, your amplifier and speakers. This would increase latency and if not done properly could reduce the sound quality.

I have found myself that GO samples per buffer should always be set to 128 - tested on many different systems over many years. And I never use jack.

csw900

**Reply by oleg68** [reply]
> how jack can "improve" sound quality

It can do it on linux by bypassing pulseaudio. Of cause, using RtAudio:alsa instead of jack would give the same result.
The only true reason for using jack it when you need to route GrandOrgue sound output to another applications (sound recorders, effect processors, OBS studio etc).

> GO samples per buffer should always be set to 128

It depends on the audio device. Some devices may work poorly with not-native buffer sizes.

**Reply by midimusic** [reply]
Thanks oleg.

Re-routing on Windows can be done without using Jack. Jack has no capability of improving sound quality - it is just a routing app which at best will not damage sound quality noticeably.

Samples per buffer is VERY non critical. Any power of 2 between 32 and 1024 will normally work. Significantly lower values will give bad sound and higher values prevent proper sound production when playing quickly. 128 is a good compromise  which works OK on every device I have ever tried and I am very doubtful that your answer is correct.

I still do not know what an xrun is, neither does google.

csw900

**Reply by oleg68** [reply]
> Significantly lower values will give bad sound

It is exactly called `xrun`.

**Comment by eturpault** [comment]
Explanation [here](https://alsa.opensrc.org/Xruns#:~:text=An%20%22xrun%22%20can%20be%20either,audible%20as%20crackles%20or%20pops.) about what a xrun is.

**Reply by thbeckpr** [reply]
Not too hard to find.

**Reply by midimusic** [reply]
Only if you already know where to look!
csw900

**Reply by midimusic** [reply]
@eturpault 
Very useful - thanks
csw900

**Comment by midimusic** [comment]
@thbeckpr 

Something which has not been mentioned above is the computer. It is evident the basic problem is that the computer is not fast enough to perform all the activities required in the time allowed i.e. data is being produced faster than it can be processed. This causes buffer overflows and bad  sound.

The cure is:
1. Get a faster computer.
2. Reduce the amount of processing needed. Not using jack will help.
3. Increase samples per buffer. (At the expense of poor performance)
4. Reduce the rate of data production - reduce sample rate to 44100 and/or reduce bits per sample.
Or any combination of the above.

csw900
